pkg/wal/wal.go

```diff
@@ -0,0 +1,345 @@
+package wal
+
+import (
+	"context"
+	"fmt"
+	iradix "github.com/hashicorp/go-immutable-radix"
+	"go.uber.org/zap"
+	"io"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/segmentio/ksuid"
+
+	"github.com/oneconcern/datamon/pkg/storage"
+)
+
+type WAL struct {
+	mutableStore       storage.Store  // Location for updating token generator object
+	tokenGeneratorPath string        	// Path to the token generator object in the mutable store
+	walStore           storage.Store  // Append only store where WAL entries are written to
+	maxConcurrency     int           	// Max concurrency when reading
+	connectionControl  chan struct{}	// How many max concurrent requests to send
+	l                  *zap.Logger    // Logging
+}
+
+type Options func(w *WAL)
+
+func MaxConcurrency(c int) Options {
+	return func(w *WAL) {
+		w.maxConcurrency = c
+	}
+}
+func TokenGeneratorPath(path string) Options {
+	return func(w *WAL) {
+		w.tokenGeneratorPath = path
+	}
+}
+
+func NewWAL(mutableStore storage.Store, walStore storage.Store, logger *zap.Logger, options ...Options) *WAL {
+	wal := WAL{
+		mutableStore: mutableStore,
+		walStore: walStore,
+		tokenGeneratorPath: tokenGeneratorPath,
+		maxConcurrency: maxConcurrency,
+		l: logger,
+	}
+	for _, option := range options {
+		option(&wal)
+	}
+	// Check if token generator object exists
+	_ = mutableStore.Put(context.Background(), wal.tokenGeneratorPath, strings.NewReader(""), storage.OverWrite)
+	return &wal
+}
+
+// Gets a token such that tokens are K-sortable.
+func (w *WAL) getToken(ctx context.Context) (string, error) {
+	err := w.updateTokenTimestamp(ctx)
```

Comment:
i think that K-sortable (aka partially sortable) is due to the total sort on time.  hence, i'm unsure why it's not necessary to, for example, acquire locks from remote. 

----

if there are two locals, A and B, at least two things can happen

- A tells the remote to look at the time (t1)
- A makes ksuid time from remote (w/ t1)
- B tells the remote to look at the time (t2)
- B makes ksuid time from remote (w/ t2)

in this case, sequential times lead to sequential ksuids

- A tells the remote to look at the time (t1)
- B tells the remote to look at the time (t2)
- A makes ksuid (k1) time from remote (w/ t2)
- B makes ksuid (k2) time from remote (w/ t2)

in this case, is there still a meaningful relation between k1 and k2?

Reply:
KSUID is a combination of a timestamp and a random string. Within a second granularity, the uniqueness depends on randomness. [KSUID how do they work](https://github.com/segmentio/ksuid#how-do-they-work)
For Datamon 2 clients uploading 2 bundles or even the same client uploading the same bundle does not have any need for strong linearization and there is no happens-before relationship.
Thus, in your example there is no relation between k1 and k2 and they would get sorted based on randomness if within the same second. For any semantic tracking of bundles, the application should depend on the labels they assign and not on the sort order. 
The current model of the latest label works if there is only one process to generate a dataset. Since there is no parent-child relationship or a single child policy for a parent between bundles, there is no real concept of HEAD as used in git. 
In the new design, the combination of pipeline run for production being controlled along with semantic versioning and schema versioning will provide the appropriate model on the consumption side for the bundles.
Will include documentation within Datamon 2.0 release for the consumption model.